### ISSUE TITLE
The SDK 'Microsoft.NET.Sdk' specified could not be found

### ISSUE THREAD

**Issue by BeardlessDev** [swiftyspiffy/TwitchLib#340]
Was following the Unity tutorial when I hit the above error while attempting to compile the DLLs.

- Installed Visual Studio 2017 via Unity, and added the .NET Framework 4.6.1 SDK and targeting pack.
- Opened up a local version of TwitchLib in VS, and got the following errors in both Debug and Release.

```
C:\Users\Fletcher\Desktop\GitLab\TwitchLib\TwitchLib\TwitchLib.csproj : error  : The SDK 'Microsoft.NET.Sdk' specified could not be found.  C:\Users\Fletcher\Desktop\GitLab\TwitchLib\TwitchLib\TwitchLib.csproj

C:\Users\Fletcher\Desktop\GitLab\TwitchLib\TwitchLib.Logging\Providers.NLog\Provider.NLog.csproj : error  : The SDK 'Microsoft.NET.Sdk' specified could not be found.  C:\Users\Fletcher\Desktop\GitLab\TwitchLib\TwitchLib.Logging\Providers.NLog\Provider.NLog.csproj

C:\Users\Fletcher\Desktop\GitLab\TwitchLib\TwitchLib.Logging\Providers.SeriLog\Provider.SeriLog.csproj : error  : The SDK 'Microsoft.NET.Sdk' specified could not be found.  C:\Users\Fletcher\Desktop\GitLab\TwitchLib\TwitchLib.Logging\Providers.SeriLog\Provider.SeriLog.csproj

C:\Users\Fletcher\Desktop\GitLab\TwitchLib\TwitchLib.Logging\Provider.Log4Net\Provider.Log4Net.csproj : error  : The SDK 'Microsoft.NET.Sdk' specified could not be found.  C:\Users\Fletcher\Desktop\GitLab\TwitchLib\TwitchLib.Logging\Provider.Log4Net\Provider.Log4Net.csproj

C:\Users\Fletcher\Desktop\GitLab\TwitchLib\TwitchLib.Testing\NetCore.Client\NetCore.Client.csproj : error  : The SDK 'Microsoft.NET.Sdk' specified could not be found.  C:\Users\Fletcher\Desktop\GitLab\TwitchLib\TwitchLib.Testing\NetCore.Client\NetCore.Client.csproj
```

In the Solution Explorer, all of the specified .csproj files had (unavailable) next to them, and when I tried to reload them, it changed to (load failed).

I clicked on the TwitchLib.csproj to see what was the problem, and was sent to this:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFrameworks>netstandard2.0;net452</TargetFrameworks>
    <PackageId>TwitchLib</PackageId>
    <Version>1.6.0</Version>
    <Description>Twitch C# library for accessing Twitch chat and whispers with events, Twitch API with both authenticated and unauthenticated calls, and PubSub with events.</Description>
    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
    <Authors>swiftyspiffy, Prom3theu5</Authors>
    <Company>swiftyspiffy (cole)</Company>
    <PackageIconUrl>https://colejelinek.com/dev/twitchlib.png</PackageIconUrl>
    <PackageProjectUrl>https://github.com/swiftyspiffy/TwitchLib</PackageProjectUrl>
    <PackageLicenseUrl>https://opensource.org/licenses/MIT</PackageLicenseUrl>
    <Copyright>Copyright 2017</Copyright>
    <PackageReleaseNotes>LivestreamMonitor changes, raid extension added, mercury field added to ChannelState, SelfRaidError, NoPermissionError and RaidedChannelIsMatureAudience events, api rate limiting, helix.GetStreams fixed, scanning for expired token, tokenexpiredexception, auth API and refresh response added for token auth, state checking on sendmessage overrides, broadcaster set correctly in GetChatters, nullreference fixed in helix videos call, rituals flag added to ChannelState, GetTopGames correctly applies parameters, support for getting and creating clips via helix endpoints, support for new auth scope clips:edit, stream up/stream down webhook subscription endpoint added, and unaccounted for event added.</PackageReleaseNotes>
    <RepositoryUrl>https://github.com/swiftyspiffy/TwitchLib</RepositoryUrl>
    <RepositoryType>Git</RepositoryType>
    <PackageTags>twitch library irc chat c# csharp api events pubsub net standard 2.0</PackageTags>
    <NeutralLanguage>en-US</NeutralLanguage>
    <AssemblyVersion>1.6.0.0</AssemblyVersion>
    <FileVersion>1.6.0.0</FileVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Newtonsoft.Json" Version="10.0.3" />
    <PackageReference Include="WebSocket4Net" Version="0.15.0" />
  </ItemGroup>

  <PropertyGroup Condition=" '$(TargetFramework)' == 'net452'">
    <DefineConstants>NET452</DefineConstants>
  </PropertyGroup>

  <PropertyGroup Condition=" '$(TargetFramework)' == 'netstandard2.0'">
    <DefineConstants>NETSTANDARD</DefineConstants>
  </PropertyGroup>
</Project>
```

After a bit of Googling, I believed that it was down to the fact I didn't have NET Standard 2.0.0 SDK and Runtime installed. I closed VS, installed both, re-opened the solution, and am still having the same issues.

I then double checked the issues list and #286 seemed to be the closest, so I followed the discussion and updated my SDK to 2.0.2, and installed the 32 versions as well, but this didn't work either.

Are there any other dependencies I need? Or steps I should have followed after installing the SDK and Runtime?

Thanks!

**Comment by swiftyspiffy** [comment]
.NET Standard shouldn't be needed to build the library as it should build in .net framework as well as .net standard/core. 
Are you able to import the NuGet package into Unity out of curiosity?

The Unity tutorial is a bit dated, but I've seen it used successfully several times.

**Comment by BeardlessDev** [comment]
When trying to import the NuGet package, I get the following error and it doesn't install.

`Failed to add reference. The package 'System.Security.Cryptography.Algorithms' tried to add a framework reference to 'mscorlib' which was not found in the GAC. This is possibly a bug in the package. Please contact the package owners for assistance.`

**Comment by Syzuna** [comment]
I actually think thats bcs its trying to install the .net standard version instead of the 4.5.2 version bcs he is using 4.6.1

**Comment by BeardlessDev** [comment]
Just installed all the targeting packs from 3.5 to 4.6.1 and tried the same thing, but with the same result both through the Unity tutorial method, and trying to install the nuget package. This time, I installed using the console and got the following:

```
Adding package 'System.Collections.Specialized.4.3.0' to folder 'C:\Users\Fletcher\Desktop\Unity\TwitchLib\packages'
Added package 'System.Collections.Specialized.4.3.0' to folder 'C:\Users\Fletcher\Desktop\Unity\TwitchLib\packages'
Install failed. Rolling back...
Package 'System.Collections.Specialized.4.3.0' does not exist in project 'TwitchLib'
Removed package 'Newtonsoft.Json.10.0.3' from 'packages.config'
Removing package 'System.Collections.Specialized.4.3.0' from folder 'C:\Users\Fletcher\Desktop\Unity\TwitchLib\packages'
Removed package 'System.Collections.Specialized.4.3.0' from folder 'C:\Users\Fletcher\Desktop\Unity\TwitchLib\packages'
Executing nuget actions took 2.48 sec
Install-Package : Failed to add reference to 'System.Collections.Specialized'.
  Project unavailable.
At line:1 char:1
+ Install-Package TwitchLib
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Install-Package], Exception
    + FullyQualifiedErrorId : NuGetCmdletUnhandledException,NuGet.PackageManagement.PowerShellCmdlets.InstallPackageCommand
```
When I tried again, on the same project, it successfully installed 'System.Collections.Specialized.4.3.0' and then I hit this:

```
Adding package 'System.Security.Cryptography.Algorithms.4.3.0' to folder 'C:\Users\Fletcher\Desktop\Unity\TwitchLib\packages'
Added package 'System.Security.Cryptography.Algorithms.4.3.0' to folder 'C:\Users\Fletcher\Desktop\Unity\TwitchLib\packages'
Install failed. Rolling back...
Package 'System.Security.Cryptography.Algorithms.4.3.0 : System.Security.Cryptography.Primitives [4.3.0, )' does not exist in project 'TwitchLib'
Removed package 'System.Security.Cryptography.Primitives.4.3.0' from 'packages.config'
Removed package 'System.Security.Cryptography.Encoding.4.3.0' from 'packages.config'
Removed package 'System.Runtime.InteropServices.4.3.0' from 'packages.config'
Removed package 'System.Runtime.Extensions.4.3.0' from 'packages.config'
Removed package 'System.Net.Sockets.4.3.0' from 'packages.config'
Removed package 'System.Net.NameResolution.4.3.0' from 'packages.config'
Removed package 'System.Linq.4.3.0' from 'packages.config'
Removed package 'System.Collections.Specialized.4.3.0' from 'packages.config'
Removed package 'Newtonsoft.Json.10.0.3' from 'packages.config'
Removing package 'System.Security.Cryptography.Algorithms.4.3.0 : System.Security.Cryptography.Primitives [4.3.0, )' from folder 'C:\Users\Fletcher\Desktop\Unity\TwitchLib\packages'
Removed package 'System.Security.Cryptography.Algorithms.4.3.0 : System.Security.Cryptography.Primitives [4.3.0, )' from folder 'C:\Users\Fletcher\Desktop\Unity\TwitchLib\packages'
Executing nuget actions took 2.66 sec
Install-Package : Failed to add reference. The package 'System.Security.Cryptography.Algorithms' tried to add a framework reference to 'mscorlib' which was not found in the GAC. This is possibly 
a bug in the package. Please contact the package owners for assistance.
  Project unavailable.
At line:1 char:1
+ Install-Package TwitchLib
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Install-Package], Exception
    + FullyQualifiedErrorId : NuGetCmdletUnhandledException,NuGet.PackageManagement.PowerShellCmdlets.InstallPackageCommand
```

Both of them start off with this:

`Attempting to gather dependency information for package 'TwitchLib.1.6.0' with respect to project 'TwitchLib', targeting '.NETFramework,Version=v4.6'`

Which I believe is a result of Unity. Unity only supports Stable .NET 3.5 Equivalent and the Experimental .NET 4.6 Equivalent, which is the one you are told to use in the Unity tutorial.

**Comment by swiftyspiffy** [comment]
From a clean slate, are you able to clone the project and build it as is without any other modifications? 
If you are unable to build the project outside of unity (just vanilla Visual Studio), then the problem is probably not with Unity, but with something else.
You might also try creating a new sample project in VS and try installing the library via NuGet and see if you can get it to reference some library component (TwitchClient?) and see if it builds.

**Comment by BeardlessDev** [comment]
... I'm sorry.

I went to create a new sample project, a Console Application, and realised I needed to install the .NET desktop development to be able to do so. Moment I installed it, all my TwitchLib issues were fixed. But hey, on the plus side, nothing is broken on your end. 👍 Appreciate the help!

Side Note: I did have to install the x32 and x64 versions of the 2.0.0+ SDK and clean the solution, but it successfully built the DLLs.
